core/src/poh_timing_reporter.rs

```diff
@@ -0,0 +1,184 @@
+use {
+    solana_metrics::{datapoint::PohTimingPoint, datapoint_info},
+    solana_sdk::clock::Slot,
+    std::{collections::HashMap, fmt},
+};
+
+/// A PohTimestamp records timing of the events during the processing of a slot
+/// by the validator
+#[derive(Debug, Clone, Copy, Default)]
+pub struct SlotPohTimestamp {
+    /// Slot start time from poh
+    pub start_time: u64,
+    /// Slot end time from poh
+    pub end_time: u64,
+    /// Last shred received time from block producer
+    pub full_time: u64,
+}
+
+/// Display trait
+impl fmt::Display for SlotPohTimestamp {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "SlotPohTimestamp: start={} end={} full={}",
+            self.start_time, self.end_time, self.full_time
+        )
+    }
+}
+
+impl SlotPohTimestamp {
+    /// Return true if the timing points of all events are received.
+    pub fn is_complete(&self) -> bool {
+        self.start_time != 0 && self.end_time != 0 && self.full_time != 0
+    }
+
+    /// Update with timing point
+    pub fn update(&mut self, timing_point: PohTimingPoint) {
+        match timing_point {
+            PohTimingPoint::PohSlotStart(ts) => self.start_time = ts,
+            PohTimingPoint::PohSlotEnd(ts) => self.end_time = ts,
+            PohTimingPoint::FullSlotReceived(ts) => self.full_time = ts,
+        }
+    }
+
+    /// Return the time difference from slot full to slot start
+    pub fn slot_full_to_start_time(&self) -> i64 {
+        (self.start_time as i64).saturating_sub(self.full_time as i64)
+    }
+
+    /// Return the time difference from slot full to slot end
+    pub fn slot_full_to_end_time(&self) -> i64 {
+        (self.end_time as i64).saturating_sub(self.full_time as i64)
+    }
+}
+
+/// A PohTimingReporter manages and reports the timing of events for incoming
+/// slots
+#[derive(Default)]
+pub struct PohTimingReporter {
+    /// Storage map of SlotPohTimestamp per slot
+    slot_timestamps: HashMap<Slot, SlotPohTimestamp>,
```

Comment:
Glad you thought about some edge cases! Namely scenarios where these slots potentially:

1. Arrive out of order (if we're jumping between forks)
2. Never complete (if we abandon a fork via poh reset, or never get the full signal)

Might be good to have some testing for these cases